### DISCUSSION TITLE
How to use esm version with CDN

### DISCUSSION THREAD

**Discussion by juliusfriedman** [vuejs/router#2276] [Help and Questions] [answered]
### Vue version

^3.4.21

### Link to minimal reproduction

https://github.com/juliusfriedman/vue-project

### Steps to reproduce

Run project

### What is expected?

Entire page loads

### What is actually happening?

<router-view> only displays html comment vnode `<!---->` where router import should be.

No error in console etc.

### System Info

```shell
All systems but only when externalizing all depencies in production mode, if I run in debug or if I remove the externalization for vue-router everything works as expected, if I instead try to internalize vue and externalize vue-router other strange issues occur.
```

### Any additional comments?

See also my question about this same problem on [StackOverflow](https://stackoverflow.com/questions/78639922/vue-3-router-issue-no-outlet-no-displayed-content?noredirect=1#comment138646499_78639922)

Thanks in advance for your help! 

If you require any additional information please let me know and I will provide ASAP!

**Comment by yangliguo7** [comment]
I cannot reproduce this issue on my end. need more information

maybe relate to ur version

**Comment by juliusfriedman** [comment]
What can't you reproduce? Locally when I build in production mode I only see the component rendered on the Home page but not the imported data of the router-view.

Are you saying when you build in production mode you are able to view the entire page and browse to home and about pages?

Let me know what other information I could possibly provide..,

Thank you!

P.s Have tried multiple versions of vue and vue-router... and removing the line of vue-router from the externals definition solves the problem but costs another 80kb

This is what I see in prod mode...
![image](https://github.com/vuejs/core/assets/7851618/b6db7aef-2441-4580-a0bd-c856cbc3c03f)

**Comment by jh-leong** [comment]
I tried removing the `external` configuration for `vue-router` in `vite.config.ts` under `rollupOptions`, and everything worked fine.

You might want to check the issues in the `vue-router` repository for more information.

<img width="778" alt="image" src="https://github.com/vuejs/core/assets/52022476/e2e64969-08cc-41b4-830e-6e759e123f11">

**Reply by juliusfriedman** [reply]
Any reason why this issue was changed into a discussion? I really would like an answer as to why this occurs.

**Reply by juliusfriedman** [reply]
Just a FYI from my debugging it seems like there is a missed check in setRef function which fails to check the owner is actually not null, It only seems to throw an error depending on the version of vue I load from the CDN...

Walking the call stack I see it skips a check because the reference is basically set like ref() which makes a ref that has no value
...

This ref is later used and then when the owner is checked to update the vnode it can't process the patch call and instead leaves a comment in the dom where it was patching

Some versions will throw an error but others will not for some reason... still looking into this... BUT it looks like a problem in vue and not vue-router... still not sure...

Also my page load gets further if I simply make a function `onMounted(()=>{})` in the earlier version but not in the later version.... its super weird.

The line in question is here:
https://github.com/vuejs/core/blob/dadb3632a0f824f6e6a1c246ed8788bea5cbcad6/packages/runtime-core/src/rendererTemplateRef.ts#L64

**Reply by kissu** [reply]
The fact that it's a discussion does not mean that it will not be fixed later on if it's indeed a bug.
Just a more open place where people can look into the future and open to discussion. 🤗

**Comment by juliusfriedman** [comment]
FWIW I tried 2 different CDNS for vue and vue-router..

https://esm.sh/
https://cdn.jsdelivr.net/npm/

Both seem to have issues where multiple versions of Vue somehow get imported...

This brings me back to a problem with the library packaging or the library itself...

Worst case we will collect scripts from vite and package them for use on our own CDN but I dont think we should have to, considering the tutorials say to use Vue and VueRouter from CDN I think this is something to investigate further.  [see](https://v3.ru.vuejs.org/guide/installation.html#cdn)

**Reply by posva** [reply, marked as answer]
I recall esm.sh having more options to properly handle dependencies.
The issue you have right now is that jsdelivr imports different versions of vue. Check the vue-router cdn file:

![Screenshot 2024-06-22 at 09 32 13](https://github.com/vuejs/router/assets/664177/12b51b44-cfc8-46fa-aafc-69924cbf58ea)


That's why you have multiple versions of Vue and why things are breaking.

Another solution would be to use import maps like the [playground](https://play.vuejs.org/#eNqNVW1v2zYQ/iucOsBKZ0uOU/SDprrpimDZsJegLfZl6gdaOltsKFIjKceB4f++I6k3O062BLape33unjtqH1SUieibDpKAVbVUhuxJroAa+FDX5EDWSlZksm1gkonWQMnGgGo1UewfB7X163S0riPvm4k+aoifi0wQEjUaQu/unyvZCBNOXqHb5CKYBq07YksNVDVH/6U1TMvL5S1wLm2y79IYH524dj940EZJsVl+bJQC0SImNTVlksatjuz35HuniNYN53eoJIeDCxP7OKmg2y7gJ4fyNybuiZHvsiDOguXPEs/kVlaQxoP+WQ+6Qknv9sE+PfFL4y5nank5jvUXgwcSt3atOo37xmC/fC/Psvk7VFI93jJt8GfaCn3cMcuzns2eT1ugS92R2glaZjvabUFHdr1kGAEpdMuGJu/I37aUva/RkUMmcWIPYfT64vVk6jWCVoAauso7SS4xpUBikx5cq6nA0ARrNsxw63SjlFR3dAOTKTl03iXjBY5F4vPbvxbDyzjOYznB09c8Uv8fUKQ/fXW/OIhfTzqmsGNj2kKHuvSMJudoDi9cLN9vPB5wyzIBO8dXAWva8C42Ds+YVrtxOlesNjhjJw4uLdZUykJjWR71Rn6RrvbwYuimQXCR3zEV1Y0uw4nfgolb965m+4314mZ2GfHhZN8Xyw4ebvvCC1eNMVKQ65yz/B43rMdwumTe8Mm2HM3nUDHRYBp7AwwbhPxiDcX563BP8BpzlEz707M7dcJnbx9iR0a6kcpqOp0GqvLSzkGLyM/ABkZtV4heCR8m+qcB9Ri1bu/fkwmC6NqOZYZeM/h6qhRgj3II98T529ltQxyInSE3nvbwH5T1DR4443QFvL3WPruYCUmZwGLIdlbJAnhkFKuQTZ8xC0hFdxzExpQoXMyRW38BdpGOOfWczCpa4y0oBbLqSstahc6CfmSzYODGirOgNKbWSRw3or7fRNjjeLC4fhNdRVdxgZs1kkagq9lKyQftrt0saJc4C67RKC5ga6TkekZr9lyKJ4bXb6O30WXM2SrG6DETBexc7K7xWKbROBBrtjkp0o4F46D+rA3DgTkqluLr8uFXJzOqgR5oXkJ+f0b+Te885DsFWN0WRsUZqnDmvPrm8x+ww3OvRBYbjtYvKD+BlryxGL3ZT40oEPbIzqH9xXHGxOaLvtkZELorygIdBtkR+fGF0ge4V9GbURe1eeSgo1zb9yW+eKc4a0x4v5VUBaiELOodQbCsIK/m8/mP7u7DcEzMVhJvlSohl/N65+Q1LQoE20swSyYwLFkSSn7Aj785vTeHNb4uxpblYpx5CD8AyPN8BCAhc/xftBGCw7+ZMEQW).

That being said, if you want to benefit from tree shaking and load less js in your application, I would recommend you not to use a CDN version for vue, vue router and pinia (a full app). Vue router is also missing the production cdn version #2242

**Reply by juliusfriedman** [reply]
Thanks for your help, I have tried inspecting the versions of the packages required manually and narrowing or setting explicitly the versions I am requesting when using externals and it still wasn't helping. 

I can change my externalization process to skip based on a package name pretty easily so perhaps I should skip `vue` and `vue-router` when externalizing...

Thanks for your help!

P.s. FWIW, I have more success with the latest vue-router@4.4.0 package!

**Reply by juliusfriedman** [reply]
@posva is it possible Vue can detect when this happens and output some type of message? It was verify confusing to me to see the patch stack mangled in ways I would not expected just from having 2 versions of the library (which seems to handle refs differently)...

A simple warning / error message `Detected another version of vue` would be great!

**Reply by posva** [reply]
I think the devtools do. It’s not possible to add a Check without having the user pass the external Vue version to do the check. Overall, it’s not worth: cdn versions are slower than bundling with libraries like Vue, Vue router, etc. you load more than you need
